pkg/project/controller/project_finalizer_controller.go

```diff
@@ -0,0 +1,138 @@
+package controller
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	kapi "k8s.io/kubernetes/pkg/api"
+
+	projectutil "github.com/openshift/origin/pkg/project/util"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+)
+
+// ProjectFinalizerController is responsible for participating in Kubernetes Namespace termination
+type ProjectFinalizerController struct {
+	client internalclientset.Interface
+
+	queue      workqueue.RateLimitingInterface
+	maxRetries int
+
+	controller cache.Controller
+	cache      cache.Store
+
+	// extracted for testing
+	syncHandler func(key string) error
+}
+
+func NewProjectFinalizerController(namespaces cache.SharedIndexInformer, client internalclientset.Interface) *ProjectFinalizerController {
+	c := &ProjectFinalizerController{
+		client:     client,
+		controller: namespaces.GetController(),
+		cache:      namespaces.GetStore(),
+		queue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		maxRetries: 10,
+	}
+	namespaces.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
```

Comment:
TODO, I think we want to generalize my naiveResourceEventHandler: https://github.com/openshift/origin/blob/master/pkg/authorization/controller/authorizationsync/generic.go#L71 .  It seems pretty common.